ArmPkg/Drivers/SmmuDxe/README.md

```diff
@@ -0,0 +1,211 @@
+
+### 2. SMMUv3 Hardware Setup
+
+The SMMU is configured in stage 2 translation mode with:
+
+- Stream table for device ID mapping
+- Command queue for SMMU operations, like TLB management
+- Event queue for error handling
+- 4KB translation granule
+
+### 3. Page Table Structure
+
+The IOMMU uses a 4-level page table structure for DMA address translation:
+<https://developer.arm.com/documentation/101811/0104/Translation-granule/The-starting-level-of-address-translation>
+
+```text
+Level 0 Table (L0)
+    ↓
+Level 1 Table (L1)
+    ↓
+Level 2 Table (L2)
+    ↓
+Level 3 Table (L3)
+    ↓
+Physical Page
+```
+
+### 4. Address Translation Process
+
+1. **Device Issues DMA**:
+   - Device uses IOVA (I/O Virtual Address)
+   - SMMU intercepts access
+
+2. **SMMU Translation**:
+   - Looks up Stream Table Entry (STE)
+   - Walks 4-level page tables
+   - Converts IOVA to PA (Physical Address)
+
+## Memory Protection
+
+The IOMMU protocol provides several protection mechanisms:
+
+1. **Access Control**:
+   - Read/Write permissions per mapping
+   - Device isolation through Stream IDs
+
+2. **Address Range Protection**:
+   - Validates DMA addresses
+   - Prevents access outside mapped regions
+
+3. **Error Handling**:
+   - Translation faults logged to Event Queue
+
+## Performance Features
+
+The implementation includes optimizations for:
+
+1. **Integration of SmmuV3 with IOMMU Protocol**
+
+2. **TLB Management**:
+   - TLB invalidation for unmapped entries
+
+## Configuration Options
+
+Key SMMU settings controlled through the SMMU config HOB:
+
+```text
+- Smmu base address, num ID's, etc.
+- Stream Table Size: Based on num IDs
+- IORT info
+```
+
+## Limitations
+
+Current implementation constraints:
+
+1. Fixed 4KB granule size
+2. 48-bit address space limit
+3. Stage 2 translation only
+4. Identity mapped page tables
+
+## Future Enhancements
+
+Potential improvements:
+
+1. Multiple translation granule support
+2. Stage 1 translation
+3. Different page table mapping schemes
+4. Updated IoMmu Protocol to optimize redundencies
+
+## Relevant Docs
+
+- SMMUv3 specification <https://developer.arm.com/documentation/ihi0070/latest/>
+- ARM a_a-profile_architecture_reference_manual <https://developer.arm.com/documentation/102105/ka-07>
+- Intel IOMMU for DMA protection in UEFI <https://www.intel.com/content/dam/develop/external/us/en/documents/intel-whitepaper-using-iommu-for-dma-protection-in-uefi.pdf>
+- IORT documentation <https://developer.arm.com/documentation/den0049/latest/>
+
+## Notes
+
+Qemu Version:
+
+- Qemu Smmu is supported on this sha - a53b931645183bd0c15dd19ae0708fc3c81ecf1d
+- Qemu emulator version 9.1.50 (v9.1.0-475-ga53b931645)
+
+TFA Version:
+
+- chery-pick:
+<https://github.com/ARM-software/arm-trusted-firmware/commit/42925c15bee09162c6dfc8c2204843ffac6201c1#diff-efe8a973d827b75aa34a8b6fd065bc9a7ffd33290d4a73697797e24e56460ae2R76>
+in Silicon/Arm/TFA - 42925c15bee09162c6dfc8c2204843ffac6201c1

```

Comment:
I would keep this doc non-Mu specific, i.e. don't mention `Silicon/ARM/TFA` and instead indicate that the linked commit is required (probably just reference the release it is in: https://github.com/ARM-software/arm-trusted-firmware/releases/tag/v2.11 or newer)

Reply:
Removing all platform specific things from this pr.

Under Qemu v9.1.50, SbsaQemuAcpiDxe.c would not run. To fix this, i had to cherry pick the mentioned change in TFA. See platform draft pr - https://github.com/microsoft/mu_tiano_platforms/pull/1085